site-src/geps/gep-726.md

```diff
@@ -0,0 +1,266 @@
+
+### Redirects
+
+The existing `RequestRedirect` filter can be expanded to support path redirects:
+
+```yaml
+kind: HTTPRoute
+apiVersion: gateway.networking.k8s.io/v1alpha2
+metadata:
+  name: http-filter-1
+spec:
+  rules:
+    - matches:
+      - path:
+          type: Prefix
+          value: /
+      filters:
+      - type: RequestRedirect
+        requestRedirect:
+          hostname: foo.com
+          path:
+            type: Prefix
+            value: /bar
+```
+
+This would be represented with the following API additions:
+```go
+// HTTPRequestRedirect defines a filter that redirects a request. At most one of
+// these filters may be used on a Route rule. This may not be used on the same
+// Route rule as a HTTPRequestRewrite filter.
+type HTTPRequestRedirect struct {
+  // Path defines a path redirect.
+  //
+  // Support: Extended
+  //
+  // +optional
+  Path *HTTPPathRedirect `json:"path,omitempty"`
+  // ...
+}
+
+// HTTPPathRedirectType defines the type of path redirect.
+type HTTPPathRedirectType string
+
+const (
+  // This type of redirect indicates that the complete path will be replaced by
+  // the path redirect value.
+  AbsoluteHTTPPathRedirect HTTPPathRedirectType = "Absolute"
+
+  // This type of redirect indicates that the prefix match will be replaced by
+  // the path redirect value. This type only applies to path matches of type
+  // Prefix. If a Route has other match types, this redirect MUST NOT be applied
+  // to them.
+  PrefixHTTPPathRedirect HTTPPathRedirectType = "Prefix"
+)
+
+// HTTPPathRedirect defines configuration for path redirects.
+type HTTPPathRedirect struct {
+  // Type defines the type of path redirect.
+  //
+  // +kubebuilder:validation:Enum=Absolute;Prefix
+  Type HTTPPathRedirectType `json:"type"`
+
+  // Value defines the HTTP path to redirect to.
+  //
+  // +kubebuilder:validation:MaxLength=1024
+  Value HTTPPathRedirectType `json:"value"`
+}
+```
+
+### Rewrites
+
+A new `RequestRewrite` filter can be added to support rewrites:
+
+```yaml
+kind: HTTPRoute
+apiVersion: gateway.networking.k8s.io/v1alpha2
+metadata:
+  name: http-filter-1
+spec:
+  rules:
+    - matches:
+      - path:
+          type: Prefix
+          value: /
+      filters:
+      - type: RequestRewrite
+        requestRewrite:
+          hostname: foo.com
+          path:
+            type: Prefix
+            value: /bar
+```
+
+This would be represent with the following API additions:
+```go
+// HTTPRequestRewrite defines a filter that modifies a request during
+// forwarding. At most one of these filters may be used on a Route rule. This
+// may not be used on the same Route rule as a HTTPRequestRedirect filter.
+type HTTPRequestRewrite struct {
+  // Hostname is the value to be used to replace the Host header value during
+  // forwarding.
+  //
+  // Support: Extended
+  //
+  // +optional
+  // +kubebuilder:validation:MaxLength=255
+  Hostname *string `json:"hostname,omitempty"`
```

Comment:
So another approach is to specify that the header filter can set the ':authority` [pseudo header](https://datatracker.ietf.org/doc/html/rfc7540#section-8.1.2.3), and the controller can program the underlying proxy to do the right thing?

Perhaps mention this in alternatives and reject it on the basis that this is more direct and explicit?

Reply:
If this is to be extended to do more than just an absolute substitution, what would that look like?